### DISCUSSION TITLE
Offsite Followup

### DISCUSSION THREAD

**Discussion by jhellerstein** [hydro-project/hydro#1203] [Ideas]
Thanks all for a successful offsite on the 15th!  Let's assemble a list of Followup plans and stakeholders. We should schedule initial meetings among the stakeholders in the coming week or two.

1. **Program Properties for Optimization**: In order to optimize programs, we need to identify the programs properties that feed into the optimization. This seems to span definitions of streams, ticks and iteration (as proposed by @shadaj), temporal persistence and operator monotonicity (as used by @davidchuyaya) and stream properties in the Hydroflow implementation like blocking/streaming/monotonicity (as implemented by @MingweiSamuel and @jhellerstein). **Stakeholders**: @shadaj, @davidchuyaya, @MingweiSamuel, @jhellerstein.
2. **Canonical Definitions**: We should develop uniform terminology for use across the team. @nacrooks specifically asked for a canonical definition of *monotonicity*. Currently we have a mix of uses across *monotone functions*, *monotonic logic*, *monotone streams* and *monotonicity over time*. Let's see clarify/contextualize/taxonomize. Are there other terms that require clarification or contextualization? **Stakeholders**: @jhellerstein @davidchuyaya @shadaj @MingweiSamuel 
3. **Engineering Practices**: We should shift some engineering practices. Some suggestions included: 
- 1-pager design docs for non-trivial projects
- git discipline: exploratory code in branches rather than main
- repo organization: explicit "nursery" for projects that are nearing relevance.
- improved doc integration: we've got some of our docs generated from docstrings. How do we improve documentation?
- inclusion of a full-time programmer in PR approval to main
- weekly demo sessions
*stakeholders*: @rohitkulshreshtha @MingweiSamuel @jhellerstein @shadaj @conor-23 @davidchuyaya

**Comment by davidchuyaya** [comment]
For monotonicity, what my rewrites care about is whether node behaves monotone across time (if the node is treated as a black box over inputs and outputs), but not necessarily that it contains monotonic functions or that facts within the node are monotone across time.

That might be important to distinguish too, because this allows for a node to be "monotonic" because it uses a threshold at the end.
On the other end of a spectrum, a 2P-Set contains monotonic logic but its output is not monotone across time.

**Reply by shadaj** [reply]
Hmm your 2P-Set distinction is interesting. Is it only not monotone if you actually compute the "true" set values at the end (by doing difference)? Like would you be able to decouple a part of the pipeline that independently transforms the +/- sides.

**Reply by davidchuyaya** [reply]
I think what would end up happening is we move the part that combines +/- sides off with state machine decoupling (it needs to process messages in the same order as its parent with the same batching). Then the monotonic parts in the middle are no longer monotonic because we need to do a count over them to find the batch size.

**Reply by jhellerstein** [reply]
Hey @davidchuyaya -- I feel like you need both monotonic ops and persistence over time. Let's consider an example with a non-monotone but `static` operator:
```
c1 = source_iter([1,2,3,4,5,6,7,8,9]) -> map(|i| i + 1);
c2 = c1 -> fold::<'static>(median) -> for_each(|i| println!("Median is {}", i));
```
As a single component, this produces exactly one value -- in the first tick. If we decouple `c1` from `c2`, the batching is different, so we are likely to see multiple printed messages, one per tick of `c2` until it gets all the data from `c1`. Depending on how we implement median, `c2` should get the right answer eventually, but it will be apparent to an observer that it's been decoupled.

I still believe what we said in the paper:
```
A simple sufficient condition for monotonicity is to ensure that (a) 𝐶2’s input relations are persisted, and (b) 𝐶2’s rules do not contain negation or aggregation
```
The operator monotonicity is in part (b).

**Reply by davidchuyaya** [reply]
What I meant was that not all functions in the node needed to be monotonic if we had a threshold at the end. Otherwise yes, non-monotonic functions with only persist would require state machine (or partial independent) decoupling, where batches and ordering is maintained.

**Reply by jhellerstein** [reply]
The idea of "downstream threshold removes the need for upstream associativity" is interesting! We should add this to the discussion of all the algebraic optimizations we can consider. 

More evidence that we may do well to break down monotonicity (for which all the ACI properties are a sufficient condition) into individual properties that we want to check.

**Reply by jhellerstein** [reply]
That said, the discussion in the paper is conservative and correct -- monotonicity of logic over time guarantees decoupling is safe. 

We do need to change your hypothesis that kicked off the discussion -- persistence over time is not enough, but perhaps  "downstream observer associativity over time" is sufficient to tolerate decoupling.

**Reply by shadaj** [reply]
I think even with thresholds though the algebraic properties are somewhat orthogonal. All you need is a partial order and a guarantee that a fold in your pipeline spits out values monotone in that partial order when given larger prefixes of the input. And the same for the boolean value that is eventually used as a condition.

**Comment by jhellerstein** [comment]
Flagging this scenario as one to explore further when we have the optimizer
conversation.
